ngraph/core/shape_inference/include/detection_output_shape_inference.hpp

```diff
@@ -0,0 +1,182 @@
+        if (num_images.is_dynamic() && class_preds_pshape[0].is_static()) {
+            num_images = class_preds_pshape[0];
+        } else {
+            NODE_VALIDATION_CHECK(
+                op,
+                class_preds_pshape[0].compatible(num_images),
+                "Class predictions' first dimension is not compatible with batch size.  Current value is: ",
+                class_preds_pshape[0],
+                ", expected: ",
+                num_images,
+                ".");
+        }
+        if (class_preds_pshape[1].is_static()) {
+            auto class_preds_pshape_2nd_dim = class_preds_pshape[1].get_length();
+            if (num_prior_boxes.is_dynamic()) {
+                NODE_VALIDATION_CHECK(op,
+                                      class_preds_pshape_2nd_dim % op->m_attrs.num_classes == 0,
+                                      "Class predictions' second dimension must be a multiply of num_classes (" +
+                                          std::to_string(op->m_attrs.num_classes) + "). Current value is: ",
+                                      class_preds_pshape_2nd_dim,
+                                      ".");
+                num_prior_boxes = class_preds_pshape_2nd_dim / op->m_attrs.num_classes;
+            } else {
+                int num_prior_boxes_val = num_prior_boxes.get_length();
+                NODE_VALIDATION_CHECK(op,
+                                      class_preds_pshape_2nd_dim == num_prior_boxes_val * op->m_attrs.num_classes,
+                                      "Class predictions' second dimension must be equal to num_prior_boxes * "
+                                      "num_classes (" +
+                                          std::to_string(num_prior_boxes_val * op->m_attrs.num_classes) +
+                                          "). Current value is: ",
+                                      class_preds_pshape_2nd_dim,
+                                      ".");
+            }
+        }
+    }
+    if (proposals_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op,
+                              proposals_pshape.size() == 3,
+                              "Proposals rank must be 3. Got " + std::to_string(proposals_pshape.size()));
+        if (num_images.is_static() && proposals_pshape[0].is_static()) {
+            int64_t proposals_1st_dim = proposals_pshape[0].get_length();
+            int64_t num_images_val = num_images.get_length();
+            NODE_VALIDATION_CHECK(op,
+                                  proposals_1st_dim == 1 || proposals_1st_dim == num_images_val,
+                                  "Proposals' first dimension is must be equal to either batch size (" +
+                                      std::to_string(num_images_val) +
+                                      ") or 1. Got: " + std::to_string(proposals_1st_dim) + ".");
+        }
+
```

Comment:
We do not need them to be static to check equality
Please correct correspondingly down below
```suggestion
            NODE_VALIDATION_CHECK(op,
                                  proposals_pshape[0].compatible(1) || proposals_pshape[0].compatible(num_images),
                                  "Proposals' first dimension is must be equal to either batch size (", num_images,
                                  ") or 1. Got: ", proposals_pshape[0], ".");
```